### ISSUE TITLE
Cache-Control has a bug on Android.

### ISSUE THREAD

**Issue by platbr** [expo/expo#1639]
Hi,

I am experiencing a strange bug on Android. It is caching the content even when "Cache-Control: no-cache" is received.
But the stranger is that debuging at expo console it doesn't show the correct Header from the api response, it shows "Cache-Control: public, max-age:0".

On iOS responseHeaders are OK, but on Android his "Cache-Control" was replaced by other one.

I used a sniffer and the correct header is sent.

### Steps to Reproduce
1. Make your APP on Android get some content from a API with header  "Cache-Control: no-cache" .
2. Turn-on airplane mode.
3. Try reload the content.

### Actual Behavior
App is using the content from cache.

### What is expected
The app should not use cache

### More info
Request (sniffer):
```
GET /api/v1/anuncios HTTP/1.1
accept: application/json, text/plain, */*
app-versao: 1.1.0
authorization: Bearer XXXXXXXXXXXXXXXXXXXXXXXXXXXXXXXXXX
Host: 172.16.3.225:4008
Connection: Keep-Alive
Accept-Encoding: gzip
User-Agent: okhttp/3.6.0
```

Response (sniffer):
```
HTTP/1.1 200 OK
X-Frame-Options: SAMEORIGIN
X-XSS-Protection: 1; mode=block
X-Content-Type-Options: nosniff
Content-Type: application/json; charset=utf-8
Cache-Control: no-cache
X-Request-Id: d9fc5763-3461-4f32-8a9d-618474b36507
X-Runtime: 3.953419
Transfer-Encoding: chunked

[{"id":97,"nome":"Evento de Testes","data":"2018-04-01T00:00:00.000-03:00","qtd_vendidos":0,"foto_url":"https://baladapp-development.s3.amazonaws.com/evento/12521/foto_evento/7d1d0ca4d50047c45d08b05fced91eb1.jpeg"}]`
```

Response object from on Android:

```
response Object {
  "config": Object {
    "adapter": [Function xhrAdapter],
    "data": undefined,
    "headers": Object {
      "APP-Versao": "1.1.0",
      "Accept": "application/json, text/plain, */*",
      "Cache-Control": "no-cache",
      "authorization": "Bearer XXXXXXXXXXXXXXXXXXXXXXXXXXXXXXXXXX",
    },
    "maxContentLength": -1,
    "method": "get",
    "timeout": 0,
    "transformRequest": Object {
      "0": [Function transformRequest],
    },
    "transformResponse": Object {
      "0": [Function transformResponse],
    },
    "url": "https://comissario-api-homologacao.baladapp.com.br/api/v1/anuncios",
    "validateStatus": [Function validateStatus],
    "xsrfCookieName": "XSRF-TOKEN",
    "xsrfHeaderName": "X-XSRF-TOKEN",
  },
  "data": Array [
    Object {
      "data": "2018-04-01T00:00:00.000-03:00",
      "foto_url": "https://baladapp-development.s3.amazonaws.com/evento/12521/foto_evento/7d1d0ca4d50047c45d08b05fced91eb1.jpeg",
      "id": 97,
      "nome": "Evento de Testes",
      "qtd_vendidos": 0,
    },
  ],
  "headers": Object {
    "cache-control": "public, max-age=0",
    "content-type": "application/json; charset=utf-8",
    "date": "Wed, 11 Apr 2018 20:16:44 GMT",
    "server": "nginx/1.13.8",
    "x-content-type-options": "nosniff",
    "x-frame-options": "SAMEORIGIN",
    "x-request-id": "d9fc5763-3461-4f32-8a9d-618474b36507",
    "x-runtime": "0.034259",
    "x-xss-protection": "1; mode=block",
  },
  "request": XMLHttpRequest {
    "DONE": 4,
    "HEADERS_RECEIVED": 2,
    "LOADING": 3,
    "OPENED": 1,
    "UNSENT": 0,
    "_aborted": false,
    "_cachedResponse": undefined,
    "_hasError": false,
    "_headers": Object {
      "accept": "application/json, text/plain, */*",
      "app-versao": "1.1.0",
      "authorization": "Bearer XXXXXXXXXXXXXXXXXXXXXXXXXXXXXXXXXX",
      "cache-control": "no-cache",
    },
    "_incrementalEvents": false,
    "_lowerCaseResponseHeaders": Object {
      "cache-control": "public, max-age=0",
      "content-type": "application/json; charset=utf-8",
      "date": "Wed, 11 Apr 2018 20:16:44 GMT",
      "server": "nginx/1.13.8",
      "x-content-type-options": "nosniff",
      "x-frame-options": "SAMEORIGIN",
      "x-request-id": "d9fc5763-3461-4f32-8a9d-618474b36507",
      "x-runtime": "0.034259",
      "x-xss-protection": "1; mode=block",
    },
    "_method": "GET",
    "_requestId": null,
    "_response": "[{\"id\":97,\"nome\":\"Evento de Testes\",\"data\":\"2018-04-01T00:00:00.000-03:00\",\"qtd_vendidos\":0,\"foto_url\":\"https://baladapp-development.s3.amazonaws.com/evento/12521/foto_evento/7d1d0ca4d50047c45d08b05fced91eb1.jpeg\"}]",
    "_responseType": "",
    "_sent": true,
    "_subscriptions": Array [],
    "_timedOut": false,
    "_trackingName": "unknown",
    "_url": "https://comissario-api-homologacao.baladapp.com.br/api/v1/anuncios",
    "readyState": 4,
    "responseHeaders": Object {
      "Cache-Control": "public, max-age=0",
      "content-type": "application/json; charset=utf-8",
      "date": "Wed, 11 Apr 2018 20:16:44 GMT",
      "server": "nginx/1.13.8",
      "x-content-type-options": "nosniff",
      "x-frame-options": "SAMEORIGIN",
      "x-request-id": "d9fc5763-3461-4f32-8a9d-618474b36507",
      "x-runtime": "0.034259",
      "x-xss-protection": "1; mode=block",
    },
    "responseURL": "https://comissario-api-homologacao.baladapp.com.br/api/v1/anuncios",
    "status": 200,
    "timeout": 0,
    "upload": XMLHttpRequestEventTarget {},
    "withCredentials": true,
  },
  "status": 200,
  "statusText": undefined,
}
```

Response object on iOS:
```
response Object {
  "config": Object {
    "adapter": [Function xhrAdapter],
    "data": undefined,
    "headers": Object {
      "APP-Versao": "1.1.0",
      "Accept": "application/json, text/plain, */*",
      "Cache-Control": "no-cache",
      "authorization": "Bearer XXXXXXXXXXXXXXXXXXXXXXXXXXXXXXXXXX",
    },
    "maxContentLength": -1,
    "method": "get",
    "timeout": 0,
    "transformRequest": Object {
      "0": [Function transformRequest],
    },
    "transformResponse": Object {
      "0": [Function transformResponse],
    },
    "url": "https://comissario-api-homologacao.baladapp.com.br/api/v1/anuncios",
    "validateStatus": [Function validateStatus],
    "xsrfCookieName": "XSRF-TOKEN",
    "xsrfHeaderName": "X-XSRF-TOKEN",
  },
  "data": Array [
    Object {
      "data": "2018-04-01T00:00:00.000-03:00",
      "foto_url": "https://baladapp-development.s3.amazonaws.com/evento/12521/foto_evento/7d1d0ca4d50047c45d08b05fced91eb1.jpeg",
      "id": 97,
      "nome": "Evento de Testes",
      "qtd_vendidos": 0,
    },
  ],
  "headers": Object {
    "cache-control": "no-cache",
    "content-encoding": "gzip",
    "content-type": "application/json; charset=utf-8",
    "date": "Wed, 11 Apr 2018 20:17:17 GMT",
    "server": "nginx/1.13.8",
    "x-content-type-options": "nosniff",
    "x-frame-options": "SAMEORIGIN",
    "x-request-id": "c738fa12-4596-45c7-ae22-47bac8cdd1e5",
    "x-runtime": "0.047903",
    "x-xss-protection": "1; mode=block",
  },
  "request": XMLHttpRequest {
    "DONE": 4,
    "HEADERS_RECEIVED": 2,
    "LOADING": 3,
    "OPENED": 1,
    "UNSENT": 0,
    "_aborted": false,
    "_cachedResponse": undefined,
    "_hasError": false,
    "_headers": Object {
      "accept": "application/json, text/plain, */*",
      "app-versao": "1.1.0",
      "authorization": "Bearer XXXXXXXXXXXXXXXXXXXXXXXXXXXXXXXXXX",
      "cache-control": "no-cache",
    },
    "_incrementalEvents": false,
    "_lowerCaseResponseHeaders": Object {
      "cache-control": "no-cache",
      "content-encoding": "gzip",
      "content-type": "application/json; charset=utf-8",
      "date": "Wed, 11 Apr 2018 20:17:17 GMT",
      "server": "nginx/1.13.8",
      "x-content-type-options": "nosniff",
      "x-frame-options": "SAMEORIGIN",
      "x-request-id": "c738fa12-4596-45c7-ae22-47bac8cdd1e5",
      "x-runtime": "0.047903",
      "x-xss-protection": "1; mode=block",
    },
    "_method": "GET",
    "_requestId": null,
    "_response": "[{\"id\":97,\"nome\":\"Evento de Testes\",\"data\":\"2018-04-01T00:00:00.000-03:00\",\"qtd_vendidos\":0,\"foto_url\":\"https://baladapp-development.s3.amazonaws.com/evento/12521/foto_evento/7d1d0ca4d50047c45d08b05fced91eb1.jpeg\"}]",
    "_responseType": "",
    "_sent": true,
    "_subscriptions": Array [],
    "_timedOut": false,
    "_trackingName": "unknown",
    "_url": "https://comissario-api-homologacao.baladapp.com.br/api/v1/anuncios",
    "readyState": 4,
    "responseHeaders": Object {
      "Cache-Control": "no-cache",
      "Content-Encoding": "gzip",
      "Content-Type": "application/json; charset=utf-8",
      "Date": "Wed, 11 Apr 2018 20:17:17 GMT",
      "Server": "nginx/1.13.8",
      "x-content-type-options": "nosniff",
      "x-frame-options": "SAMEORIGIN",
      "x-request-id": "c738fa12-4596-45c7-ae22-47bac8cdd1e5",
      "x-runtime": "0.047903",
      "x-xss-protection": "1; mode=block",
    },
    "responseURL": "https://comissario-api-homologacao.baladapp.com.br/api/v1/anuncios",
    "status": 200,
    "timeout": 0,
    "upload": XMLHttpRequestEventTarget {
      Symbol(listeners): Object {},
    },
    "withCredentials": true,
    Symbol(listeners): Object {
      "error": Object {
        "kind": 3,
        "listener": [Function handleError],
        "next": null,
      },
      "readystatechange": Object {
        "kind": 3,
        "listener": [Function handleLoad],
        "next": null,
      },
      "timeout": Object {
        "kind": 3,
        "listener": [Function handleTimeout],
        "next": null,
      },
    },
  },
  "status": 200,
  "statusText": undefined,
}

iOS
```
  
### Environment

  OS: macOS High Sierra 10.13.3
  Node: 8.4.0
  Yarn: 1.5.1
  npm: 5.3.0
  Watchman: Not Found
  Xcode: Xcode 9.3 Build version 9E145
  Android Studio: Not Found

Packages: (wanted => installed)
  expo: ^26.0.0 => 26.0.0
  react: 16.3.0-alpha.1 => 16.3.0-alpha.1
  react-native: https://github.com/expo/react-native/archive/sdk-26.0.0.tar.gz => 0.54.2

Diagnostics report:
  https://exp-xde-diagnostics.s3.amazonaws.com/baladapp-a4f3bc14-a54e-4eda-9c11-d2ea4f6129b9.tar.gz

**Comment by schazers** [comment]
Thanks for the report! We'll take a look :)

**Comment by esamelson** [comment]
Hi @platbr - sorry for the delay here. Thanks for the detailed report. It looks like this is likely due to this upstream issue: https://github.com/facebook/react-native/issues/16795 . We'll see what we can do but just want to let you know we may not be able to fix this as soon as you'd like.

**Comment by platbr** [comment]
@esamelson tks =)

**Comment by cmeredith** [comment]
Upvoted. This is a pretty significant security issue. @esamelson -- is there any workaround where we can at least inject functionality to prevent caching?

**Comment by ide** [comment]
You could add a random cache-busting query param to every URL in order to tell the client not to cache. My understanding is that the server isn't sending back a `Cache-Control: public` header so there's no risk of (spec-compliant) intermediate servers (ex: proxies) caching the response; the only intermediary undesirably caching the response is OkHttp on the device.

**Comment by esamelson** [comment]
It looks like this has been fixed upstream in RN 0.60.2 and above, so the fix should be included in SDK 36 which will use RN 0.61.x.

**Comment by esamelson** [comment]
Closing since SDK 36 is out with RN 0.61.4.
